File: src/deriver.rb
require_relative 'browser'
class Deriver
  :browser
  :location
  BASE_URL = 'https://www.google.com/maps/'

  def initialize(x, y) 
    @browser = Browser.new 
    @location = { :x => x, :y => y }
  end
  
  def start
    @browser.maximize
    @browser.go_to url
    sleep 5
    @browser.drag_pegman
    sleep 3
    @browser.click
    sleep 3
    self
  end

  def run 
    while true 
      @browser.send_key :up
      sleep 1
      @browser.send_key left_or_right
      sleep 1
    end
    self 
  end

  def fin() @browser.fin end

  private def url() BASE_URL + "@%{x},%{y},15z" % @location end
  private def choose() rand(2) == 1 end
  private def left_or_right() choose ? :right : :left end
end

File: app/src/main/kotlin/src/Deriver.kt
package src

const val BASE_URL = "https://www.google.com/maps/"

data class Location(val x: Double, val y: Double)

class Deriver {
  private val browser: Browser
  private var location: Location
  private val isStuck: Boolean
    get() = browser.location?.let { it.x == location.x && it.y == location.y } ?: true

  constructor(x: Double, y: Double) {
    location = Location(x, y)
    browser = Browser(getUrl(location))
  }

  fun walk(): Unit {
    browser.forward()
    getUnstuck()
    browser.location?.let { location = it }
  }

  fun getUnstuck(): Unit {
    val turn = if (Math.random() > 0.5) browser::left else browser::right

    while (isStuck) {
      browser.forward()
      turn()
      browser.forward()
    }
  }

  fun finish(): Unit = browser.finish()

  private fun getUrl(loc: Location): String {
    val x = loc.x
    val y = loc.y
    return BASE_URL + "@$x,$y,15z"
  }
}
